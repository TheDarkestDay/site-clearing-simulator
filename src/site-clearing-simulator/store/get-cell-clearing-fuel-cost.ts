import { SiteCellType } from './store-slice';

export const getCellClearingFuelCost = (cellType: SiteCellType): number => {
  switch (cellType) {
    case 'o':
    case 'C':
      return 1;
    case 'r':
    case 't':
      return 2;
    default:
      throw new Error('Clearing of a cell with bulldozer or preservable tree is not allowed');
  }
};