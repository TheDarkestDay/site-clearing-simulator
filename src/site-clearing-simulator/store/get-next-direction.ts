import { BulldozerDirection } from './store-slice';

export const getNextDirection = (direction: BulldozerDirection, turnDirection: 'left' | 'right'): BulldozerDirection => {
  const directions = [
    BulldozerDirection.Up,
    BulldozerDirection.Right,
    BulldozerDirection.Down,
    BulldozerDirection.Left,
  ];

  const directionDelta = turnDirection === 'right' ? 1 : -1;
  const newDirectionIndex = directions.indexOf(direction) + directionDelta;

  if (newDirectionIndex < 0) {
    return BulldozerDirection.Left;
  }

  if (newDirectionIndex >= directions.length) {
    return BulldozerDirection.Up;
  }

  return directions[newDirectionIndex];
}