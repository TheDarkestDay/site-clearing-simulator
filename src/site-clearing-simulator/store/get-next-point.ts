import { Point2D } from './point-2d';
import { BulldozerDirection } from './store-slice';

export const getNextPoint = ({ x, y }: Point2D, direction: BulldozerDirection): Point2D => {
  switch (direction) {
    case BulldozerDirection.Up:
      return {
        x: x - 1,
        y,
      }
    case BulldozerDirection.Right:
      return {
        x,
        y: y + 1,
      }
    case BulldozerDirection.Down:
      return {
        x: x + 1,
        y
      }
    case BulldozerDirection.Left:
      return {
        x,
        y: y - 1
      }
  }
};