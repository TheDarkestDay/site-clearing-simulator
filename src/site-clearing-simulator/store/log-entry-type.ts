export enum LogEntryType {
  RotateLeft = 'ROTATE_LEFT',

  RotateRight = 'ROTATE_RIGHT',

  MoveTo = 'MOVE_TO',

  Stop = 'STOP',
}