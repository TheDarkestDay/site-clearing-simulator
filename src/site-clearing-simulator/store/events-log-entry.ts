import { LogEntryType } from './log-entry-type';

export type EventsLogEntry = {
  type: LogEntryType;
  description: string;
};