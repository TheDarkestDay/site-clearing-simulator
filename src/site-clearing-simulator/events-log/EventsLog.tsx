import { Card, List, ListItem, ListItemIcon, ListItemText, Typography } from '@material-ui/core'
import { NavigationOutlined, RedoOutlined, Stop, UndoOutlined } from '@material-ui/icons';
import React, { useLayoutEffect, useRef } from 'react';
import { useSelector } from 'react-redux';
import { LogEntryType } from '../store/log-entry-type';
import { getEvents } from '../store/selectors';
import styles from './styles.module.css';

type Props = {
  className?: string;
}

const getLogEntryIcon = (type: LogEntryType) => {
  switch (type) {
    case LogEntryType.MoveTo:
      return <NavigationOutlined />;
    case LogEntryType.RotateLeft:
      return <UndoOutlined />;
    case LogEntryType.RotateRight:
      return <RedoOutlined />;
    case LogEntryType.Stop:
      return <Stop />;
  }
};

export const EventsLog = ({ className }: Props) => {
  const events = useSelector(getEvents);
  const listRef = useRef<HTMLUListElement>(null);

  useLayoutEffect(() => {
    const listElement = listRef.current;
    if (!listElement) {
      return;
    }

    listElement.scrollTop = listElement.offsetHeight;
  });

  return (
    <Card className={className}>
      <Typography component="h2" className={styles.logTitle}>
        Log
      </Typography>

      <List ref={listRef} className={styles.logList} dense>
        {events.map((event, index) => <ListItem key={index}>
          <ListItemIcon>
            {getLogEntryIcon(event.type)}
          </ListItemIcon>

          <ListItemText primary={event.description} />
        </ListItem>)}
      </List>
    </Card>
  );
};