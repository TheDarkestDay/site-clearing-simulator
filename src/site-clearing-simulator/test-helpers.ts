import { SiteClearingSimulatorState, BulldozerDirection } from './store/store-slice';

export const getTestState = (
  overrides: Partial<SiteClearingSimulatorState> = {}
): SiteClearingSimulatorState => {
  return {
    map: [],
    isStarted: false,
    isStopped: false,
    bulldozerDirection: BulldozerDirection.Right,
    bulldozerPosition: {
      x: 0,
      y: -1,
    },
    isPreservedTreeRemoved: false,
    fuelUsed: 0,
    eventsLog: [],
    ...overrides,
  };
};