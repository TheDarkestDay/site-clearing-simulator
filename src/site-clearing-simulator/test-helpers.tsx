import { createStore } from '@reduxjs/toolkit';
import { create } from 'react-test-renderer';
import React from 'react';
import { Provider } from 'react-redux';
import { GlobalState, rootReducer } from '../store';
import { SiteClearingSimulatorState, BulldozerDirection, initialState } from './store/store-slice';

export const getTestState = (
  overrides: Partial<SiteClearingSimulatorState> = {}
): SiteClearingSimulatorState => {
  return {
    map: [],
    isStarted: false,
    isStopped: false,
    bulldozerDirection: BulldozerDirection.Right,
    bulldozerPosition: {
      x: 0,
      y: -1,
    },
    isPreservedTreeRemoved: false,
    fuelUsed: 0,
    eventsLog: [],
    ...overrides,
  };
};

export const renderWithStore = (component: React.ReactNode, state: Partial<GlobalState['siteClearingSimulator']> = {}) => {
  const testStore = createStore(rootReducer, {
    siteClearingSimulator: {
      ...initialState,
      ...state,
    }
  });
  const consumer = (
    <Provider store={testStore}>
      {component}
    </Provider>
  );

  return create(consumer);
}