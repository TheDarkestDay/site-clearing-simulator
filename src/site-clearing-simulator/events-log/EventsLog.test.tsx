import { LogEntryType } from '../store/log-entry-type';
import { renderWithStore } from '../test-helpers';
import { EventsLog } from './EventsLog';

describe('EventsLog component', () => {
  it('should match snapshot', () => {
    const wrapper = renderWithStore(<EventsLog />, {
      eventsLog: [
        {
          type: LogEntryType.RotateLeft,
          description: 'Rotate left'
        },
        {
          type: LogEntryType.RotateRight,
          description: 'Rotate left'
        }, {
          type: LogEntryType.MoveTo,
          description: 'Move to 1-1',
        },
        {
          type: LogEntryType.Stop,
          description: 'Simulation was stopped',
        }
      ]
    });

    expect(wrapper).toMatchSnapshot();
  });
});