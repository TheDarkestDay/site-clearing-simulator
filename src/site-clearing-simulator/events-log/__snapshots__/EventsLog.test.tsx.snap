// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsLog component should match snapshot 1`] = `
<div
  className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <h2
    className="MuiTypography-root logTitle MuiTypography-body1"
  >
    Log
  </h2>
  <ul
    className="MuiList-root logList MuiList-dense MuiList-padding"
  >
    <li
      className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemIcon-root"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12.5 8c-2.65 0-5.05.99-6.9 2.6L2 7v9h9l-3.62-3.62c1.39-1.16 3.16-1.88 5.12-1.88 3.54 0 6.55 2.31 7.6 5.5l2.37-.78C21.08 11.03 17.15 8 12.5 8z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          Rotate left
        </span>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemIcon-root"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.4 10.6C16.55 8.99 14.15 8 11.5 8c-4.65 0-8.58 3.03-9.96 7.22L3.9 16c1.05-3.19 4.05-5.5 7.6-5.5 1.95 0 3.73.72 5.12 1.88L13 16h9V7l-3.6 3.6z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          Rotate left
        </span>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemIcon-root"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 7.27l4.28 10.43-3.47-1.53-.81-.36-.81.36-3.47 1.53L12 7.27M12 2L4.5 20.29l.71.71L12 18l6.79 3 .71-.71L12 2z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          Move to 1-1
        </span>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemIcon-root"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 6h12v12H6z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root MuiListItemText-dense"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
        >
          Simulation was stopped
        </span>
      </div>
    </li>
  </ul>
</div>
`;
